launchpad-guaranteed-tickets-v2/src/guaranteed_ticket_winners.rs

```diff
@@ -0,0 +1,205 @@
+    random::Random,
+    tickets::{TicketRange, WINNING_TICKET},
+};
+use multiversx_sc::api::CryptoApi;
+
+const VEC_MAPPER_START_INDEX: usize = 1;
+
+#[derive(TopEncode, TopDecode, NestedEncode, NestedDecode)]
+pub struct GuaranteedTicketsSelectionOperation<M: ManagedTypeApi + CryptoApi> {
+    pub rng: Random<M>,
+    pub leftover_tickets: usize,
+    pub leftover_ticket_pos_offset: usize,
+    pub total_additional_winning_tickets: usize,
+}
+
+impl<M: ManagedTypeApi + CryptoApi> Default for GuaranteedTicketsSelectionOperation<M> {
+    fn default() -> Self {
+        Self {
+            rng: Random::default(),
+            leftover_tickets: 0,
+            leftover_ticket_pos_offset: 1,
+            total_additional_winning_tickets: 0,
+        }
+    }
+}
+
+pub enum AdditionalSelectionTryResult {
+    Ok,
+    CurrentAlreadyWinning,
+    NewlySelectedAlreadyWinning,
+}
+
+#[multiversx_sc::module]
+pub trait GuaranteedTicketWinnersModule:
+    launchpad_common::launch_stage::LaunchStageModule
+    + launchpad_common::config::ConfigModule
+    + launchpad_common::ongoing_operation::OngoingOperationModule
+    + launchpad_common::tickets::TicketsModule
+    + crate::guaranteed_tickets_init::GuaranteedTicketsInitModule
+{
+    fn select_guaranteed_tickets(
+        &self,
+        op: &mut GuaranteedTicketsSelectionOperation<Self::Api>,
+    ) -> OperationCompletionStatus {
+        let mut users_whitelist = self.users_with_guaranteed_ticket();
+        let mut users_left = users_whitelist.len();
+
+        self.run_while_it_has_gas(|| {
+            if users_left == 0 {
+                return STOP_OP;
+            }
+
+            let current_user = users_whitelist.get_by_index(VEC_MAPPER_START_INDEX);
+            let _ = users_whitelist.swap_remove(&current_user);
+            users_left -= 1;
+
+            let user_ticket_status_mapper = self.user_ticket_status(&current_user);
+            if user_ticket_status_mapper.is_empty() {
+                return CONTINUE_OP;
+            }
+            let user_confirmed_tickets = self.nr_confirmed_tickets(&current_user).get();
+            let user_ticket_status = user_ticket_status_mapper.get();
+
+            let mut user_guaranteed_tickets_no = 0;
+            let mut user_leftover_tickets = 0;
+
+            for info in user_ticket_status.guaranteed_tickets_info.iter() {
+                if user_confirmed_tickets >= info.min_confirmed_tickets {
+                    user_guaranteed_tickets_no += info.guaranteed_tickets;
+                } else {
+                    user_leftover_tickets += info.guaranteed_tickets;
+                }
+            }
+
+            if user_guaranteed_tickets_no > user_confirmed_tickets {
+                let excess_tickets = user_guaranteed_tickets_no - user_confirmed_tickets;
+                user_leftover_tickets += excess_tickets;
+                user_guaranteed_tickets_no = user_confirmed_tickets;
+            }
```

Comment:
I don't get this logic at all. Care to elaborate?

Reply:
The idea was to have an array of guaranteed tickets, each with its own needed confirmation number. Eg. 1 from EGLD staking (64 confirmed tickets requirement), 1 for energy (32 confirmed) and 3 for having a premium account (3 confirmed). Then, the next step would be, what happens if the user has confirmed only 10 tickets. He should only receive 3 guaranteed, as the threshold was not reached for the other 2.